confluent_kafka/schema_registry/json_schema.py

```diff
@@ -0,0 +1,294 @@
+        elsewhere in this document, is not strictly required by the JSON Schema
+        specification. It is however required by this Serializer. This
+        annotation(record name) is used to register the Schema with the Schema
+        Registry. See documentation below for additional details on Subjects
+        and schema registration.
+
+    Args:
+        schema_registry_client (SchemaRegistryClient): Schema Registry
+            client instance.
+
+        schema_str (str): JSON Schema definition.
+
+        to_dict (callable, optional): Callable(SerializationContext, object) -> dict.
+            Converts object to a dict.
+
+        conf (dict): JsonSerializer configuration.
+
+    .. _Subject name strategy:
+        https://docs.confluent.io/current/schema-registry/serializer-formatter.html#subject-name-strategy
+
+    .. _Schema definition:
+        https://json-schema.org/understanding-json-schema/reference/generic.html
+
+    """  # noqa: E501
+    __slots__ = ['_hash', '_auto_register', '_known_subjects', '_parsed_schema',
+                 '_registry', '_schema', '_schema_id', '_schema_name',
+                 '_subject_name_func', '_to_dict']
+
+    # default configuration
+    _default_conf = {'auto.register.schemas': True,
+                     'subject.name.strategy': topic_subject_name_strategy}
+
+    def __init__(self, schema_registry_client, schema_str, to_dict=None,
+                 conf=None):
+        self._registry = schema_registry_client
+        self._schema_id = None
+        # Avoid calling registry if schema is known to be registered
+        self._known_subjects = set()
+
+        if to_dict is not None and not callable(to_dict):
+            raise ValueError("to_dict must be callable with the signature"
+                             " to_dict(SerializationContext, object)->dict")
+
+        self._to_dict = to_dict
+
+        # handle configuration
+        conf_copy = self._default_conf.copy()
+        if conf is not None:
+            conf_copy.update(conf)
+
+        self._auto_register = conf_copy.pop('auto.register.schemas')
+        if not isinstance(self._auto_register, bool):
+            raise ValueError("auto.register.schemas must be a boolean value")
+
+        self._subject_name_func = conf_copy.pop('subject.name.strategy')
+        if not callable(self._subject_name_func):
+            raise ValueError("subject.name.strategy must be callable")
+
+        if len(conf_copy) > 0:
+            raise ValueError("Unrecognized properties: {}"
+                             .format(", ".join(conf_copy.keys())))
+
+        schema_dict = json.loads(schema_str)
+        schema_name = schema_dict.get('title', None)
+        if schema_name is None:
+            raise ValueError("Missing required JSON schema annotation title")
+
+        self._schema_name = schema_name
+        self._parsed_schema = schema_dict
+        self._schema = Schema(schema_str, schema_type="JSON")
+
+    def __call__(self, ctx, obj):
```

Comment:
it makes more sense to me for ctx to be last (and to match .net). if you changed this on behalf of a comment i made, i apologize, didn't mean it here. i meant ctx should be first in the subject naming case.

Reply:
reasoning: here the extensibility you're most likely to want is going to be inside SerializationContext. In the subject naming case, the extensibility you might care about is outside it.

Reply:
sigh,  yes I wanted them to be consistent across the board so I changed both. The serializers unfortunately being the more difficult one to change. Alright I'll fix it here then open a PR for generic serdes and another for avro